### ISSUE TITLE
Should "Run exit node" GUI checkboxes also prevent computer from sleeping?

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#3574]
Tailscale 1.20 on both Windows and Mac will support a "Run exit node" checkbox.

I bet the first support requests we get will be people losing access to them when their desktops at home go to sleep.

Should running an exit node (or subnet router) prevent the computer from sleeping?

Should that be a separate checkbox?

Should we just warn people about them and have them use a separate app (like [Caffeine](https://www.macupdate.com/app/mac/24120/caffeine) etc)?

/cc @crawshaw @DentonGentry @mayakacz @tailscale/frontend

**Comment by crawshaw** [comment]
I don't think it should be a separate checkbox. If you run as an exit node, sleep should be disabled. But I would be fine with shipping the first version without that, just with a warning.

**Comment by sonovawolf** [comment]
Yeah, having an app that prevents your computer from sleeping but is not purpose-built for that effect is really unexpected, despite making total sense for this one feature. I would definitely give a warning, whether we are preventing sleep or not.

**Comment by bradfitz** [comment]
Then next questions:

* How do we prevent macOS from sleeping? From all [three variants](https://tailscale.com/kb/1065/macos-variants/). @nickoneill?
* How do we prevent Windows from sleeping? @dblohm7?

**Comment by crawshaw** [comment]
On macOS, I/O kit can do it but I have no idea if we can use it from inside the sandbox. https://developer.apple.com/library/archive/qa/qa1340/_index.html

**Comment by nickoneill** [comment]
Definitely possible from the sandbox, there are a few Mac App Store apps that provide this functionality. I'll have to take a look at various approaches to see if it is something that can be shared between all variants. My gut says `tailscaled` may need to take a different approach.

**Comment by gabefair** [comment]
And is there a way to listen for sleep requests from the OS? 

This way a notification from tailscale could say `tailscale is preventing this machine from sleeping while in use as an exit node`

But also, I can think of times where I have tailscale in use as an exit node on my field laptop (for my home network to see what my laptop sees) but my laptop to sleep when I close the lid.

**Comment by gabefair** [comment]
Or there should just be a small note under the use as exit-node checkbox reminding the user to disable sleep on host OS.

This way tailscale doesn't have to police the OS's settings. Leaving tailscale to worry about tailscale

**Comment by d1vanloon** [comment]
PowerToys Awake may be a helpful reference for Windows: https://github.com/microsoft/PowerToys/tree/main/src/modules/awake

**Comment by vwkd** [comment]
It'd be great if Tailscale could prevent a machine from sleeping if there is any meaningful incoming traffic (other than keep-alive pings) more generally, not necessarily only when used as exit node. This is useful when one wants to connect through Tailnet to the machine, e.g. SSH, SMB, etc.

At least on macOS, currently incoming Tailscale traffic doesn't seem to prevent macOS from sleeping and cutting of any incoming SSH, SMB connections through Tailscale. This is different if the incoming connections come from the LAN which prevent macOS from sleeping.

So, one needs at least one other always-on machine in the LAN through which one can relay the Tailscale traffic into the LAN or proxy through with a traditional Wireguard VPN connection like a router.

**Comment by gabefair** [comment]
Unlike startup apps, most OSs don't have a GUI method to view apps that are setup to block sleeping/shutdown. This is why if this feature is to be implemented, the OS would need to tell the user which app (tailscale) is blocking it from doing its job.
